src/main/java/org/alfresco/repo/security/permissions/impl/acegi/ACLEntryAfterInvocationProvider.java

```diff
@@ -573,244 +554,194 @@ private ResultSet decideNew(Authentication authentication, Object object, Config
+                        else if (cad.typeString.equals(AFTER_ACL_PARENT))
+                        {
+                            testNodeRef = returnedObject.getChildAssocRef(i).getParentRef();
+                        }
+
+                        if(isUnfiltered(testNodeRef))
+                        {
+                            continue;
+                        }
+
+                        if (filteringResultSet.getIncluded(i) && (testNodeRef != null) && (permissionService.hasPermission(testNodeRef, cad.required.toString()) == AccessStatus.DENIED))
+                        {
+                            filteringResultSet.setIncluded(i, false);
+                        }
+                    }
+                }
+            }
         }
         finally
         {
-           // put them back to how they were
-           returnedObject.setBulkFetch(oldBulkFetch);
-           returnedObject.setBulkFetchSize(oldFetchSize);
+            // put them back to how they were
+            returnedObject.setBulkFetch(oldBulkFetch);
+            returnedObject.setBulkFetchSize(oldFetchSize);
         }
-        
+
         return filteringResultSet;
     }
 
-
-    private ResultSet decideOld(Authentication authentication, Object object, ConfigAttributeDefinition config, ResultSet returnedObject) throws AccessDeniedException
-
+    /**
+     * Compute a (Weak)FilteringResultSet by selecting the first maxSize elements from returnedObject.
+     *
+     * @param maxSize
+     * @param returnedObject
+     * @return
+     */
+    private ResultSet filterMaxCount(Integer maxSize, ResultSet returnedObject)
     {
-        if (returnedObject == null)
+        WeakFilteringResultSet filteringResultSet = new WeakFilteringResultSet(returnedObject);
+        if (maxSize == null)
         {
-            return null;
+            return returnedObject;
         }
+        else
+        {
+            for (int i = 0; i < maxSize.intValue() && i < returnedObject.length(); i++)
+            {
+                filteringResultSet.setIncluded(i, true);
+            }
 
-        FilteringResultSet filteringResultSet = new FilteringResultSet(returnedObject);
-
-        List<ConfigAttributeDefintion> supportedDefinitions = extractSupportedDefinitions(config);
+            if (returnedObject.length() > maxSize.intValue())
+            {
+                filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.FINAL_SIZE, PermissionEvaluationMode.EAGER, returnedObject.getResultSetMetaData()
+                        .getSearchParameters()));
+            }
+            else
+            {
+                filteringResultSet.setResultSetMetaData(new SimpleResultSetMetaData(LimitBy.UNLIMITED, PermissionEvaluationMode.EAGER, returnedObject.getResultSetMetaData()
+                        .getSearchParameters()));
+            }
+        }
+        return filteringResultSet;
+    }
 
+    /**
+     * Get the mac size from a resultSet.

```

Comment:
Also - it would be nice to expand on what "max size" really means.  I think this is an upper bound on the number of results that could be included in the result set?  What would cause the number of results to be less than the output from this method?